File: Reference/tests/Brontide.Reference.Studio.Tests/ComponentProviderRestartAttemptRecoveryTests.cs
using System.Text.Json;
using Brontide.Reference.Experimental.ComponentManagement;

namespace Brontide.Reference.Studio.Tests;

public sealed class ComponentProviderRestartAttemptRecoveryTests
{
    private static readonly ProviderRestartAttemptRunId RunId = ProviderRestartAttemptRunId.Create("restart-run.test.1");
    private static readonly OccurrenceId Occurrence = OccurrenceId.Create("occ.def.test.cooling-provider.1");
    private static readonly ProviderArtifactSetId Staged = ProviderArtifactSetId.Create(new string('A', 64));
    private static readonly ProviderRestartPolicy Policy = ProviderRestartPolicy.Create(2, TimeSpan.FromMinutes(1));
    private static readonly DateTimeOffset Start = new(2026, 8, 5, 9, 0, 0, TimeSpan.Zero);

    private sealed class TemporaryJournal : IDisposable
    {
        public TemporaryJournal() => Path = System.IO.Path.Combine(System.IO.Path.GetTempPath(), $"brontide-cbi53-{Guid.NewGuid():N}", "restart.journal");
        public string Path { get; }
        public void Dispose()
        {
            var root = System.IO.Path.GetDirectoryName(Path)!;
            if (Directory.Exists(root)) Directory.Delete(root, recursive: true);
        }
    }

    private static DurableProviderRestartAttemptJournal Establish(string path) =>
        DurableProviderRestartAttemptJournal.Establish(path, RunId, Occurrence, Staged, Policy).Journal!;

    private static ProviderRestartAttemptJournalTransitionResult Commit(
        DurableProviderRestartAttemptJournal journal, DateTimeOffset instant, string code)
    {
        Assert.That(journal.BeginAttempt(instant).Code, Is.EqualTo("durable-restart-attempt-started"));
        var completed = code == "provider-restart-completed";
        return journal.CommitAttempt(
            code,
            completed ? "none" : code == "portable-process-interrupted" ? "cbi2" : "cbi31",
            providerStarted: code == "portable-process-interrupted" || completed,
            lifecycleReconstructed: completed,
            completed);
    }

    [Test]
    public void Cbi53_C1_one_journal_names_one_bounded_restart_lineage()
    {
        using var temporary = new TemporaryJournal();
        var established = DurableProviderRestartAttemptJournal.Establish(temporary.Path, RunId, Occurrence, Staged, Policy);
        var duplicate = DurableProviderRestartAttemptJournal.Establish(temporary.Path, RunId, Occurrence, Staged, Policy);
        var mismatch = DurableProviderRestartAttemptJournal.Open(
            temporary.Path, RunId, OccurrenceId.Create("occ.def.test.other.1"), Staged);
        Assert.Multiple(() =>
        {
            Assert.That(established.Code, Is.EqualTo("durable-restart-established"));
            Assert.That(duplicate.Code, Is.EqualTo("durable-restart-already-exists"));
            Assert.That(mismatch.Code, Is.EqualTo("durable-restart-lineage-mismatch"));
            Assert.That(established.Journal!.Snapshot.MaximumAttempts, Is.EqualTo(2));
        });
    }

    [Test]
    public void Cbi53_C2_every_transition_is_atomic_and_integrity_checked()
    {
        using var temporary = new TemporaryJournal();
        var journal = Establish(temporary.Path);
        var original = File.ReadAllBytes(temporary.Path);
        Directory.CreateDirectory(temporary.Path + ".tmp");
        var refused = journal.BeginAttempt(Start);
        Directory.Delete(temporary.Path + ".tmp");
        Assert.Multiple(() =>
        {
            Assert.That(refused.Code, Is.EqualTo("durable-restart-write-failed"));
            Assert.That(File.ReadAllBytes(temporary.Path), Is.EqualTo(original));
        });
        var bytes = File.ReadAllBytes(temporary.Path);
        bytes[0] ^= 0x7F;
        File.WriteAllBytes(temporary.Path, bytes);
        Assert.That(DurableProviderRestartAttemptJournal.Open(temporary.Path, RunId, Occurrence, Staged).Code,
            Is.EqualTo("durable-restart-corrupt"));
    }

    [Test]
    public void Cbi53_C3_non_ready_policy_history_changes_no_journal_state()
    {
        using var temporary = new TemporaryJournal();
        var journal = Establish(temporary.Path);
        Commit(journal, Start, "staged-artifact-integrity-failed");
        var before = File.ReadAllBytes(temporary.Path);
        var waiting = journal.BeginAttempt(Start.AddSeconds(59));
        Assert.Multiple(() =>
        {
            Assert.That(waiting.Code, Is.EqualTo("durable-restart-waiting"));
            Assert.That(File.ReadAllBytes(temporary.Path), Is.EqualTo(before));
        });
    }

    [Test]
    public void Cbi53_C4_in_flight_state_precedes_restart_effects()
    {
        using var temporary = new TemporaryJournal();
        var journal = Establish(temporary.Path);
        Assert.That(journal.BeginAttempt(Start).Code, Is.EqualTo("durable-restart-attempt-started"));
        var reopened = DurableProviderRestartAttemptJournal.Open(temporary.Path, RunId, Occurrence, Staged);
        Assert.Multiple(() =>
        {
            Assert.That(reopened.Code, Is.EqualTo("durable-restart-indeterminate"));
            Assert.That(reopened.Journal!.Snapshot.InFlightIndex, Is.Zero);
        });
    }

    [Test]
    public void Cbi53_C5_committed_failures_drive_delay_and_exhaustion()
    {
        using var temporary = new TemporaryJournal();
        var journal = Establish(temporary.Path);
        Commit(journal, Start, "portable-process-interrupted");
        var exhausted = Commit(journal, Start.AddMinutes(1), "staged-artifact-integrity-failed");
        Assert.Multiple(() =>
        {
            Assert.That(exhausted.Code, Is.EqualTo("durable-restart-exhausted"));
            Assert.That(exhausted.Snapshot.Attempts, Has.Count.EqualTo(2));
        });
    }

    [Test]
    public void Cbi53_C6_interrupted_work_requires_explicit_reconciliation()
    {
        using var retryFile = new TemporaryJournal();
        var retry = Establish(retryFile.Path);
        retry.BeginAttempt(Start);
        var ready = retry.ResolveInterrupted(ProviderRestartAttemptRecoveryDecision.Retry);
        using var abandonFile = new TemporaryJournal();
        var abandon = Establish(abandonFile.Path);
        abandon.BeginAttempt(Start);
        var abandoned = abandon.ResolveInterrupted(ProviderRestartAttemptRecoveryDecision.Abandon);
        Assert.Multiple(() =>
        {
            Assert.That(ready.Code, Is.EqualTo("durable-restart-retry-ready"));
            Assert.That(ready.Snapshot.RetryCount, Is.EqualTo(1));
            Assert.That(abandoned.Code, Is.EqualTo("durable-restart-abandoned"));
        });
    }

    [Test]
    public void Cbi53_C7_terminal_recovery_is_idempotent_and_effect_free()
    {
        using var temporary = new TemporaryJournal();
        var journal = Establish(temporary.Path);
        Commit(journal, Start, "provider-restart-completed");
        var before = File.ReadAllBytes(temporary.Path);
        var reopened = DurableProviderRestartAttemptJournal.Open(temporary.Path, RunId, Occurrence, Staged).Journal!;
        var after = reopened.BeginAttempt(Start.AddMinutes(1));
        Assert.Multiple(() =>
        {
            Assert.That(after.Code, Is.EqualTo("durable-restart-completed"));
            Assert.That(File.ReadAllBytes(temporary.Path), Is.EqualTo(before));
        });
    }

    [Test]
    public void Cbi53_C8_reference_executes_the_shared_history_model()
    {
        using var fixture = JsonDocument.Parse(File.ReadAllText(System.IO.Path.Combine(
            TestContext.CurrentContext.TestDirectory, "component-management", "fixtures",
            "cbi53-durable-restart-attempt-vectors.json")));
        foreach (var vector in fixture.RootElement.GetProperty("vectors").EnumerateArray())
        {
            using var temporary = new TemporaryJournal();
            var journal = Establish(temporary.Path);
            var now = Start;
            var code = "durable-restart-established";
            foreach (var actionElement in vector.GetProperty("actions").EnumerateArray())
            {
                var action = actionElement.GetString()!;
                if (action.StartsWith("attempt:", StringComparison.Ordinal))
                    code = Commit(journal, now, action[8..]).Code;
                else if (action.StartsWith("advance:", StringComparison.Ordinal))
                    now = now.AddSeconds(int.Parse(action[8..], System.Globalization.CultureInfo.InvariantCulture));
                else if (action == "crash") code = journal.BeginAttempt(now).Code;
                else if (action == "reopen")
                {
                    var opened = DurableProviderRestartAttemptJournal.Open(temporary.Path, RunId, Occurrence, Staged);
                    code = opened.Code;
                    journal = opened.Journal!;
                }
                else if (action == "retry") code = journal.ResolveInterrupted(ProviderRestartAttemptRecoveryDecision.Retry).Code;
                else if (action == "abandon") code = journal.ResolveInterrupted(ProviderRestartAttemptRecoveryDecision.Abandon).Code;
            }
            var expectedInFlight = vector.GetProperty("expectedInFlight");
            Assert.Multiple(() =>
            {
                Assert.That(code, Is.EqualTo(vector.GetProperty("expectedCode").GetString()), vector.GetProperty("name").GetString());
                Assert.That(journal.Snapshot.Phase, Is.EqualTo(vector.GetProperty("expectedPhase").GetString()));
                Assert.That(journal.Snapshot.Attempts.Select(item => item.Code), Is.EqualTo(vector.GetProperty("expectedAttemptCodes").EnumerateArray().Select(item => item.GetString())));
                Assert.That(journal.Snapshot.NextAttemptIndex, Is.EqualTo(vector.GetProperty("expectedNextAttempt").GetInt32()));
                Assert.That(journal.Snapshot.InFlightIndex, Is.EqualTo(expectedInFlight.ValueKind == JsonValueKind.Null ? null : expectedInFlight.GetInt32()));
                Assert.That(journal.Snapshot.InterruptionCount, Is.EqualTo(vector.GetProperty("expectedInterruptions").GetInt32()));
                Assert.That(journal.Snapshot.RetryCount, Is.EqualTo(vector.GetProperty("expectedRetries").GetInt32()));
            });
        }
    }
}

File: Minimal/tests/Brontide.Minimal.Host.Tests/ComponentProviderRestartAttemptRecoveryTests.fs
namespace Brontide.Minimal.Host.Tests

open System
open System.Globalization
open System.IO
open System.Text.Json
open NUnit.Framework
open Brontide.Minimal.Experimental.ComponentManagement
open Brontide.Minimal.Host

type private RestartTemporaryJournal() =
    let path = Path.Combine(Path.GetTempPath(), $"brontide-cbi53-{Guid.NewGuid():N}", "restart.journal")
    member _.Path = path
    interface IDisposable with
        member _.Dispose() =
            match Path.GetDirectoryName path with
            | null | "" -> ()
            | root when Directory.Exists root -> Directory.Delete(root, true)
            | _ -> ()

[<TestFixture>]
type ComponentProviderRestartAttemptRecoveryTests() =
    let multiple action = Assert.Multiple(Action action)
    let runIdentity = ProviderRestartAttemptRunId.create "restart-run.test.1"
    let occurrence = OccurrenceId.create "occ.def.test.cooling-provider.1"
    let staged = ProviderArtifactSetId.create (String('A', 64))
    let policy = ProviderRestartPolicy.create 2 (TimeSpan.FromMinutes 1.0)
    let start = DateTimeOffset(2026, 8, 5, 9, 0, 0, TimeSpan.Zero)

    let establish path =
        DurableProviderRestartAttemptJournal.Establish(path, runIdentity, occurrence, staged, policy).Journal.Value

    let commit (journal: DurableProviderRestartAttemptJournal) instant code =
        Assert.That(journal.BeginAttempt(instant).Code, Is.EqualTo "durable-restart-attempt-started")
        let completed = code = "provider-restart-completed"
        let refusedBy = if completed then "none" elif code = "portable-process-interrupted" then "cbi2" else "cbi31"
        journal.CommitAttempt(code, refusedBy, code = "portable-process-interrupted" || completed, completed, completed)

    [<Test>]
    member _.``CBI53 C1 one journal names one bounded restart lineage``() =
        use temporary = new RestartTemporaryJournal()
        let established = DurableProviderRestartAttemptJournal.Establish(temporary.Path, runIdentity, occurrence, staged, policy)
        let duplicate = DurableProviderRestartAttemptJournal.Establish(temporary.Path, runIdentity, occurrence, staged, policy)
        let mismatch = DurableProviderRestartAttemptJournal.Open(temporary.Path, runIdentity, OccurrenceId.create "occ.def.test.other.1", staged)
        multiple (fun () ->
            Assert.That(established.Code, Is.EqualTo "durable-restart-established")
            Assert.That(duplicate.Code, Is.EqualTo "durable-restart-already-exists")
            Assert.That(mismatch.Code, Is.EqualTo "durable-restart-lineage-mismatch")
            Assert.That(established.Journal.Value.Snapshot.MaximumAttempts, Is.EqualTo 2))

    [<Test>]
    member _.``CBI53 C2 every transition is atomic and integrity checked``() =
        use temporary = new RestartTemporaryJournal()
        let journal = establish temporary.Path
        let original = File.ReadAllBytes temporary.Path
        Directory.CreateDirectory(temporary.Path + ".tmp") |> ignore
        let refused = journal.BeginAttempt start
        Directory.Delete(temporary.Path + ".tmp")
        multiple (fun () ->
            Assert.That(refused.Code, Is.EqualTo "durable-restart-write-failed")
            Assert.That(File.ReadAllBytes temporary.Path, Is.EqualTo(box original)))
        let bytes = File.ReadAllBytes temporary.Path
        bytes[0] <- bytes[0] ^^^ 0x7Fuy
        File.WriteAllBytes(temporary.Path, bytes)
        Assert.That(DurableProviderRestartAttemptJournal.Open(temporary.Path, runIdentity, occurrence, staged).Code,
            Is.EqualTo "durable-restart-corrupt")

    [<Test>]
    member _.``CBI53 C3 non ready policy history changes no journal state``() =
        use temporary = new RestartTemporaryJournal()
        let journal = establish temporary.Path
        commit journal start "staged-artifact-integrity-failed" |> ignore
        let before = File.ReadAllBytes temporary.Path
        let waiting = journal.BeginAttempt(start.AddSeconds 59.0)
        multiple (fun () ->
            Assert.That(waiting.Code, Is.EqualTo "durable-restart-waiting")
            Assert.That(File.ReadAllBytes temporary.Path, Is.EqualTo(box before)))

    [<Test>]
    member _.``CBI53 C4 in flight state precedes restart effects``() =
        use temporary = new RestartTemporaryJournal()
        let journal = establish temporary.Path
        Assert.That(journal.BeginAttempt(start).Code, Is.EqualTo "durable-restart-attempt-started")
        let reopened = DurableProviderRestartAttemptJournal.Open(temporary.Path, runIdentity, occurrence, staged)
        multiple (fun () ->
            Assert.That(reopened.Code, Is.EqualTo "durable-restart-indeterminate")
            Assert.That(reopened.Journal.Value.Snapshot.InFlightIndex, Is.EqualTo(Some 0)))

    [<Test>]
    member _.``CBI53 C5 committed failures drive delay and exhaustion``() =
        use temporary = new RestartTemporaryJournal()
        let journal = establish temporary.Path
        commit journal start "portable-process-interrupted" |> ignore
        let exhausted = commit journal (start.AddMinutes 1.0) "staged-artifact-integrity-failed"
        multiple (fun () ->
            Assert.That(exhausted.Code, Is.EqualTo "durable-restart-exhausted")
            Assert.That(exhausted.Snapshot.Attempts, Has.Length.EqualTo 2))

    [<Test>]
    member _.``CBI53 C6 interrupted work requires explicit reconciliation``() =
        use retryFile = new RestartTemporaryJournal()
        let retry = establish retryFile.Path
        retry.BeginAttempt(start) |> ignore
        let ready = retry.ResolveInterrupted ProviderRestartAttemptRecoveryDecision.Retry
        use abandonFile = new RestartTemporaryJournal()
        let abandon = establish abandonFile.Path
        abandon.BeginAttempt(start) |> ignore
        let abandoned = abandon.ResolveInterrupted ProviderRestartAttemptRecoveryDecision.Abandon
        multiple (fun () ->
            Assert.That(ready.Code, Is.EqualTo "durable-restart-retry-ready")
            Assert.That(ready.Snapshot.RetryCount, Is.EqualTo 1)
            Assert.That(abandoned.Code, Is.EqualTo "durable-restart-abandoned"))

    [<Test>]
    member _.``CBI53 C7 terminal recovery is idempotent and effect free``() =
        use temporary = new RestartTemporaryJournal()
        let journal = establish temporary.Path
        commit journal start "provider-restart-completed" |> ignore
        let before = File.ReadAllBytes temporary.Path
        let reopened = DurableProviderRestartAttemptJournal.Open(temporary.Path, runIdentity, occurrence, staged).Journal.Value
        let after = reopened.BeginAttempt(start.AddMinutes 1.0)
        multiple (fun () ->
            Assert.That(after.Code, Is.EqualTo "durable-restart-completed")
            Assert.That(File.ReadAllBytes temporary.Path, Is.EqualTo(box before)))

    [<Test>]
    member _.``CBI53 C8 minimal executes the shared history model``() =
        use fixture = JsonDocument.Parse(File.ReadAllText(Path.Combine(
            TestContext.CurrentContext.TestDirectory, "component-management", "fixtures",
            "cbi53-durable-restart-attempt-vectors.json")))
        let text (value: JsonElement) = value.GetString() |> Option.ofObj |> Option.defaultValue ""
        for vector in fixture.RootElement.GetProperty("vectors").EnumerateArray() do
            use temporary = new RestartTemporaryJournal()
            let mutable journal = establish temporary.Path
            let mutable now = start
            let mutable code = "durable-restart-established"
            for actionElement in vector.GetProperty("actions").EnumerateArray() do
                let action = text actionElement
                if action.StartsWith("attempt:", StringComparison.Ordinal) then
                    code <- (commit journal now action[8..]).Code
                elif action.StartsWith("advance:", StringComparison.Ordinal) then
                    now <- now.AddSeconds(Double.Parse(action[8..], CultureInfo.InvariantCulture))
                elif action = "crash" then code <- journal.BeginAttempt(now).Code
                elif action = "reopen" then
                    let opened = DurableProviderRestartAttemptJournal.Open(temporary.Path, runIdentity, occurrence, staged)
                    code <- opened.Code
                    journal <- opened.Journal.Value
                elif action = "retry" then code <- journal.ResolveInterrupted(ProviderRestartAttemptRecoveryDecision.Retry).Code
                elif action = "abandon" then code <- journal.ResolveInterrupted(ProviderRestartAttemptRecoveryDecision.Abandon).Code
            let expectedInFlight = vector.GetProperty "expectedInFlight"
            multiple (fun () ->
                Assert.That(code, Is.EqualTo(vector.GetProperty("expectedCode") |> text), vector.GetProperty("name") |> text)
                Assert.That(journal.Snapshot.Phase, Is.EqualTo(vector.GetProperty("expectedPhase") |> text))
                Assert.That(journal.Snapshot.Attempts |> List.map _.Code,
                    Is.EqualTo(box (vector.GetProperty("expectedAttemptCodes").EnumerateArray() |> Seq.map text |> Seq.toList)))
                Assert.That(journal.Snapshot.NextAttemptIndex, Is.EqualTo(vector.GetProperty("expectedNextAttempt").GetInt32()))
                Assert.That(journal.Snapshot.InFlightIndex,
                    Is.EqualTo(box (if expectedInFlight.ValueKind = JsonValueKind.Null then None else Some(expectedInFlight.GetInt32()))))
                Assert.That(journal.Snapshot.InterruptionCount, Is.EqualTo(vector.GetProperty("expectedInterruptions").GetInt32()))
                Assert.That(journal.Snapshot.RetryCount, Is.EqualTo(vector.GetProperty("expectedRetries").GetInt32())))
